### PR TITLE
fs-util: versioned directory support

### PR DESCRIPTION
**Pull request by topimiettinen**
In case the name of the parent directory for `RootImage=` or `RootDirectory=` ends
with `auto.d` suffix, the path is searched for items starting with the path but
using versioned suffixes. The highest version is selected.

Examples:

`RootImage=` exists in file system:
```bash
$ systemctl cat lvm2-monitor.service | grep RootImage=
RootImage=/usr/local/lib/rootimages.auto.d/lvm-all.squashfs
$ ls -1d /usr/local/lib/rootimages.auto.d/lvm-all.squashfs*
/usr/local/lib/rootimages.auto.d/lvm-all.squashfs-a
/usr/local/lib/rootimages.auto.d/lvm-all.squashfs-b
```

`RootImage=` is a LVM volume in VG `service.auto.d`:
```bash
$ systemctl cat acpid.service | grep RootImage=
RootImage=/dev/service.auto.d/%p-all.squashfs
$ ls -1d /dev/service.auto.d/acpid-all.squashfs*
/dev/service.auto.d/acpid-all.squashfs-a
/dev/service.auto.d/acpid-all.squashfs-b
```

`RootDirectory=` exists in file system:
```bash
$ systemctl cat systemd-udevd.service | grep RootDirectory=
RootDirectory=/usr/local/lib/rootimages.auto.d/systemd-udevd-%v.dir
$ ls -1d /usr/local/lib/rootimages.auto.d/systemd-udevd-*
/usr/local/lib/rootimages.auto.d/systemd-udevd-5.10.0-8-amd64.dir-a
/usr/local/lib/rootimages.auto.d/systemd-udevd-5.10.0-8-amd64.dir-b
/usr/local/lib/rootimages.auto.d/systemd-udevd-5.14.0-2-amd64.dir-a
/usr/local/lib/rootimages.auto.d/systemd-udevd-5.14.0-2-amd64.dir-b
```

`systemd-nspawn`:
```bash
$ ls -1d /tmp/foo.auto.d/debian*
/tmp/foo.auto.d/debian-a
/tmp/foo.auto.d/debian-b
$ SYSTEMD_LOG_LEVEL=debug systemd-nspawn -D /tmp/foo.auto.d/debian -U --machine debian
...
Trying to version /tmp/foo.auto.d/debian.
Found possible version /tmp/foo.auto.d/debian-a.
Found possible version /tmp/foo.auto.d/debian-b.
Versioned path from /tmp/foo.auto.d/debian to /tmp/foo.auto.d/debian-b.
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Spawning container debian on /tmp/foo.auto.d/debian-b.
```

### CODE REVIEW

**Comment by topimiettinen** [comment]
CI fails and I forgot to document nspawn.

**Comment by topimiettinen** [comment]
Fedora fails with bogus error:
```
../src/core/namespace.c: In function â€˜setup_namespaceâ€™:
../src/fundamental/macro-fundamental.h:275:17: error: â€˜root_imageâ€™ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  275 |                 free(memory);                   \
      |                 ^~~~
../src/core/namespace.c:1909:30: note: â€˜root_imageâ€™ was declared here
 1909 |         _cleanup_free_ char *root_image;
      |                              ^~~~~~~~~~
cc1: all warnings being treated as errors
```
The variable is unconditionally initialized on the next line. But let's please the compiler and add useless initializers.

**Comment by topimiettinen** [comment]
Only focal-ppc64el and focal-s390x fail, probably not related:
```
upstream             FAIL timed out
```

**Comment by topimiettinen** [comment]
@poettering PTAL.

File: man/systemd-nspawn.xml
Comment (poettering) [inline man/systemd-nspawn.xml]: please say `auto.d/`, i.e. suffix the `/`, to clarify it's a directory. We try to do this in all our docs: suffix dirs with a slash.

File: man/systemd-nspawn.xml
Comment (poettering) [inline man/systemd-nspawn.xml]: highest → newest

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c]: please don't use globs for this. globs are OK if users actually provide us with a glob expression. But if we use a constant our even generated glob expression just use opendir()/readdir() insteads, it's simpler and faster.

Comment (poettering) [reply src/basic/fs-util.c]: also, I think we should skip any dot prefixed entries, and filter out entries of type we are not interested in (i.e. if we are called for S_IFDIR ignore anything that is not S_IFDIR, and if we are called for S_IFREG similar). (But mybe allow symlinks in either case)

Comment (poettering) [reply src/basic/fs-util.c]: Finally, I'd leave a door open for later extensions: for sd-boot we reserve entry names starting in a `@` to have special meanings: we don't allow them to appear on disk, but they are resolved to special meaning in different contexts. I would keep that door open here too. Hence please ignore any entries starting with `@` and debug log about them, that they are reserved (to make this debuggable). That way we can be sure not to create compat problems should we later on give them meaning.

(what do i mean by all this: specifically, we could eventually allow users to define a symlink `@default` that can point to any entry, and that allows users to override the version picked entry and make some explicitly picked version the one to use).

e.g:

```
foo.auto.d/0.1/
foo.auto.d/0.2/
foo.auto.d/0.3/
foo.auto.d/0.4/
foo.auto.d/@default → 0.3
```

in such a setup 0.4 would of course be the entry to boot if nothing is specified. But if the `@default` symlink exists we'd allow users to override this behaviour that way, and (temporarily?) pick a different one (i.e. 0.3 in this case)

Now, I am not saying we should implement this immediately (though if you like, go ahead, I think it makes sense), but I'd definitely want to keep the door open for this, and hence for now ignore all `@` prefixed entries.

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c]: please propagate any errors we see via the return value. i.e. we typically prefer that the caller ignores errors when that's a good idea, not the callee.

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c]: hmm, what I had in mind was that we could do this:

```
systemd-nspawn --image=/srv/myimages.auto.d/ -b
```

And that we'd then pick the newest image file inside that dir. Or:

```
systemd-nspawn --directory=/srv/myimages.auto.d -b
```

i.e. we'd specify the dir to look into. But the code above takes the *parent* of the specified dir, and that's what I don't grok?

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c]: this will fall appart if people do this:

```
systemd-nspawn --directory='/srv/my***images.auto.d' -b
```

i.e. people have a path with an asterisk, and they explicitly quoted it so that the asterisks aren't understood as globs. Yet here they will be interpreted nonetheless. Hence: please do not use globs for internal enumerations.

File: src/nspawn/nspawn.c
Comment (poettering) [inline src/nspawn/nspawn.c]: oom checks please.

Comment (poettering) [reply src/nspawn/nspawn.c]: also please do so only if --image= or --directory= are specified, and only for the parameter of that switch. please do not do this if we automatically found a machine image with --machine=/-M.

**Review by poettering — CHANGES_REQUESTED** [review]

File: man/systemd-nspawn.xml
Comment (topimiettinen) [reply man/systemd-nspawn.xml]: OK, though "newest" could be interpreted as "newest mtime", but that's not the case, especially for devices.

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: Makes sense

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: That would work too. So if the path is S_IFDIR, all files would be checked, but otherwise only items which match the path, using the final path component as a prefix. With the final path component, it's possible to put all RootImages etc. into same folder.

I want to use `foo.auto.d/systemd-networkd` as the path (matching only for example `foo.auto.d/systemd-networkd-1.0`, `foo.auto.d/systemd-networkd-2.0` etc.) as that's much simpler with LVM, otherwise I'd have to create a new VG for each service and that would kind of defeat the LVM idea.

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: Ouch. That should apply to all cases where the path comes from untrusted source and there's some kind of trust boundary? Luckily there doesn't seem to be any CVEs related to [glob](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=glob), but could you be the first one to notice this?

File: src/basic/fs-util.c
Comment (poettering) [reply src/basic/fs-util.c]: I don't understand what LVM has to do with this?

Comment (topimiettinen) [reply src/basic/fs-util.c]: I'm using LVM (when possible) for the RootImages. I don't think it's possible to make subdirectories (nested VGs) in LVM (something like `/dev/services/systemd-networkd.auto.d/` with LVs `ver-{1,2}`, instead of `/dev/services.auto.d/` containing all LVs ($SERVICE-{1,2}) for all services), so I can't have one `.auto.d/` per service without constructing new VGs for each service (`/dev/systemd-networkd.auto.d/`) which I don't want. Or maybe it is by stacking LVs over LVs, I have to check.

File: man/systemd-nspawn.xml
Comment (topimiettinen) [reply man/systemd-nspawn.xml]: fixed

File: man/systemd-nspawn.xml
Comment (topimiettinen) [reply man/systemd-nspawn.xml]: fixed as "newest"

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: Last version doesn't use glob() but readdir(). I've added filename checks including `@`.

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: Also directory without filename prefix works now

File: src/nspawn/nspawn.c
Comment (topimiettinen) [reply src/nspawn/nspawn.c]: OOM checks added. I didn't test machine part but I moved the calls earlier.

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: Not using globs anymore

File: src/basic/fs-util.c
Comment (topimiettinen) [reply src/basic/fs-util.c]: Changed the signature and logic for this

**Comment by topimiettinen** [comment]
Focal-ppc64el and focal-s390x fail is most probably unrelated: 
```
upstream             FAIL timed out
```

**Comment by topimiettinen** [comment]
Maybe `.auto.d/` concept could apply nicely also to MountImages and ExtensionImages. For example, the binaries of a service might be stored in a base image with `RootImage=`, but the configuration files of /etc in a `MountImage=`. The administrator may want to upgrade the base and configuration images with loosely coupled cycles. PID1 would automatically select the newest image for both and in the future, optionally downgrade to previous version.

I was also thinking about how to version service configuration files (/etc/systemd/system/xyz.service etc., so something like /etc/systemd/system/xyz.service.auto.d/ or /etc/systemd/system.auto.d/xyz.service*), but maybe that could get messy.

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1032]: this should have a test case in src/test/test-fs-util.c

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1035]: merge these three lines?

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1047]: plese don't stat first and open then. Things like that have this TOCTTOU smell to it. Instead, just opendir() the thing. If this works, you can just go ahead with the logic after checking the the path ends in auto.d. OTOH, if this fails with ENOTDIR you can just skip the whole process, as apparently you dealing with an existing file that is not a directory. If it fails with ENOENT then look for the parent, and do your prefix match.

Comment (poettering) [reply src/basic/fs-util.c:1047]: (or in other words: the return info from opendir() already tells you enough information)

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1060]: stat() returns errors in errno. thus `return -errno` here. Or alternatively use our new RET_NERRNO() construct

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1070]: we do error handling with *negative* errno errors i.e. do `return -errno`

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1074]: consider using readdir_no_dot()

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1087]: if you use readdir_no_dot() (or readdir_ensure_type()) you can rely on d_type and thus drop DT_UNKNOWN

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1101]: free_and_strdup()

File: src/core/namespace.c
Comment (poettering) [inline src/core/namespace.c:1883]: merge these two lines?

File: src/core/namespace.c
Comment (poettering) [inline src/core/namespace.c:1911]: please add a comment `/* want_dir= */` before the `false` to make clear what the argument is about. `false` alone is otherwise so opaque

File: src/nspawn/nspawn.c
Comment (poettering) [inline src/nspawn/nspawn.c:2993]: we use the `arg_` prefix so far only for global variables that are directly process parameters (i.e. parsed from the cmdline). Please name these local variables differently to avoid any confusion.

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1126]: I don#t think we should do a memory allocation here. Just return > 0 in case we did the versioning thing or == 0 if we didn't and then initialize `*ret_path = NULL`. The caller can then look at the return value and if it's > 0 use the return and otherwise ignore it...

File: src/basic/fs-util.c
Comment (poettering) [inline src/basic/fs-util.c:1097]: merge this if block with the next, really. the body of the if block is pretty much the same...

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
> Maybe `.auto.d/` concept could apply nicely also to MountImages and ExtensionImages. For example, the binaries of a service might be stored in a base image with `RootImage=`, but the configuration files of /etc in a `MountImage=`. The administrator may want to upgrade the base and configuration images with loosely coupled cycles. PID1 would automatically select the newest image for both and in the future, optionally downgrade to previous version.

Yes, extending this to MountImages= makes a ton of sense to me.

i.e. I think adding versioning like this makes sense for "big" objects, i.e. disk images, OS trees, service trees and such. But I'd avoid it for "small" objects. i.e. individual service files.

I wonder if ".auto.d/" is really the best name for this? maybe call it `.version.d"` or `.autoversion.d` or `.pick.d` or `.vpick.d` or so?

btw, @keszybz what's your take on this idea? naming suggestion?

**Comment by poettering** [comment]
I have been thinking about the semantics a bit more, and I think I am not a fan of the logic currently implemented in the PR that there basically two things you can specify: either path to a dir, or a path to a dir suffixed by a prefix.

I understand you want the prefix behaviour, and I think I can accept that (though I find the LVM usecase questionnable, not a fan of LVM and its semantics here)...

So hence, maybe we should drop the "auto.d/" identifier, and instead focus on the prefix thing with a single indicator the anable this behaviour. Long story short, I think i'd propose this:

if a path is suffixed by three dots `...` we'll enable this behaviour, and that's the whole rule. Example:

```
systemd-nspawn --image=/some/dir/...
```

That would do the version logic over the full names in the `/some/dir/` directory. Or to cover your use case:

```
systemd-nspawn --image=/some/dir/some-prefix-...
```

Which would then enuemrate the contents of /some/dir/ and do the version logic over all entries starting with `some-prefix`.

I think that's a simpler approach, and paths on disk that end in `...` are hopefully not something that happens IRL so that we can get away with redefining what paths written that way mean.

I think I like this approach much better, and it's a lot simpler. What do you think?

**Comment by topimiettinen** [comment]
I like `...` too. Using a known wildcard character like `*` would work and it would be maybe even more familiar to users, but then someone would want to use full globbing, regexps and whatever.

**Comment by poettering** [comment]
Yeah, I'd stay away with stuff that already has meaning elsewhere. But afaics "..." is not used by anything relevant so far, so I think we are good if we use it.

**Comment by poettering** [comment]
Thinking further though: I figure in many cases a ".raw" suffix should be matched too. Hence maybe allow the "..." to appear in the middle of an fname too:

```
systemd-nspawn --image=/some/path/myprefix....raw
```

which then slowly but sure becomes something that is awfully close to the MatchPattern= logic #20156 uses.

**Comment by topimiettinen** [comment]
$v would be OK too, no need to use different symbols for same idea.

**Comment by topimiettinen** [comment]
I hadn't followed the sysupdate PR closely, but it looks like `pattern_match()` would be 95% to 150% of what's needed here. So maybe put this on hold until #20156 is finalized.

**Comment by bluca** [comment]
More food for thoughts (nothing to implement here and now ofc): a natural extension of the A/B schemes is automatic fallback. You switch to a newer version, it fails, but there's an older version? Try to use that one automatically.

This is a common pattern - we are using it right now with some custom logic by the orchestrator we use. But that requires special handling when the orchestrator itself is updated. Being able to integrate the A/B scheme with automatic fallback into `portablectl reattach` would be super cool.

**Comment by topimiettinen** [comment]
> More food for thoughts (nothing to implement here and now ofc): a natural extension of the A/B schemes is automatic fallback. You switch to a newer version, it fails, but there's an older version? Try to use that one automatically.

Yes, that's the plan. Maybe it needs to be opt-in with new directive like `StartAutoDowngrade=yes` or something. For example, if `StartLimitBurst=5`, with `StartAutoDowngrade=yes`, 4 start attempts could be used for newest version and 1 for fallback to previous. It should be relatively simple but the command to switch from newest to previous needs to be passed from unit.c (where the attempt counter is kept) down to namespace.c (or get count from namespace.c or some other ugly layering violation).

**Comment by poettering** [comment]
Hmm, so if we want assessment counting like sd-boot implements it, then maybe we should indeed reuse the match pattern stuff from sysupdate eventually, because it has support for all that.

So let's let's get sysupdate merged first? once that's done, let's move its match pattern logic into src/shared/ and reuse it here?

**Comment by topimiettinen** [comment]
> Hmm, so if we want assessment counting like sd-boot implements it, then maybe we should indeed reuse the match pattern stuff from sysupdate eventually, because it has support for all that.
> 
> So let's let's get sysupdate merged first? once that's done, let's move its match pattern logic into src/shared/ and reuse it here?

Yes. I'm not sure how useful this PR would be then as a basis, maybe just close this.

Another idea: if there are lots (hundreds) of versions of an image (for example, automatically generated) and it's noticed that there's an error under seldomly used conditions, it might be useful to be able to bisect quickly the version that caused the issue. This could be implemented using the version mechanisms. It's probably better handled by an external bisection tool though, but that tool would then need a way to force a specific version.

**Comment by poettering** [comment]
I guess we can close this now that 76511c1bd32a262c76d462919083925c47cbd212 has been merged.

**Pull request closed without merging**
